lib/statistics/statistical_test/chi_squared_test.rb

```diff
@@ -37,6 +37,59 @@ def self.goodness_of_fit(alpha, expected, observed)
           alternative: p_value <= alpha,
           confidence_level: 1 - alpha }
       end
+
+      # The following three functions serve to calculate a test of independence for contingency
+      # tables of the type
+      #
+      #      A   B
+      # X   20  18
+      # Y    7  35
+      #
+      # They have been tested using 2x2 and 3x3 tables.
+      #
+      def self.test_of_independence(alpha, observed)
+        expected = calculate_expected(observed)
+        df = (observed.size - 1) * (observed.transpose.size - 1)
+        chi_score = calculate_chi_score(observed, expected)
+        probability = 1.0 - Statistics::Distribution::ChiSquared.new(df).cumulative_function(chi_score)
+        p_value = 1.0 - probability
+    
+        {
+          chi_score: chi_score,
+          df: df,
+          probability: probability,
+          p_value: p_value,
+          alpha: alpha,
+          #null: alpha < p_value,
+          #alternative: p_value <= alpha,
+          confidence_level: 1 - alpha,
+          expected: expected
+        }
+      end
+      
+      # For a contingency table of observed values, calculate the expected values
+      def self.calculate_expected(observed)
+        row_sums = observed.map { |row| row.sum.to_f }
+        col_sums = observed.transpose.map { |col| col.sum.to_f }
+        total_sum = row_sums.sum
+    
+        expected = observed.map.with_index { |row, i|
+          row.map.with_index { |obs, j|
+            (row_sums[i] * col_sums[j]) / total_sum
+          }
+        }
```

Comment:
so for ruby syntax, I prefer to use do/end block notation instead of brackets for multi lines blocks. It makes everything a bit more clear IMHO.

Reply:
you can also do something like this instead:
```ruby
observed.zip(expected).map { |observed, expected| (observed * expected) / total_sum.to_r }
```

notice the `to_r` that tries to convert the number to a rational one. The float is giving me lots of issues with calculations and accuracy with big numbers.

Reply:
I (locally) added the `do ... end`, but the solution with `map` didn't work :( Thanks for the hint with the `to_r`, will build that into other calculations as well.